pkg/networkserver/grpc_deviceregistry.go

```diff
@@ -45,82 +46,206 @@ func (ns *NetworkServer) Set(ctx context.Context, req *ttnpb.SetEndDeviceRequest
 	if err := rights.RequireApplication(ctx, req.Device.ApplicationIdentifiers, ttnpb.RIGHT_APPLICATION_DEVICES_WRITE); err != nil {
 		return nil, err
 	}
+
+	gets := append(req.FieldMask.Paths[:0:0], req.FieldMask.Paths...)
+	if ttnpb.HasAnyField(req.FieldMask.Paths, "queued_application_downlinks") {
+		gets = append(gets,
+			"mac_state.device_class",
+		)
+	}
+	if ttnpb.HasAnyField(req.FieldMask.Paths, "mac_state.device_class") {
+		gets = append(gets,
+			"mac_state.current_parameters",
+			"mac_state.desired_parameters",
+			"queued_application_downlinks",
+		)
+	}
+
 	var addDownlinkTask bool
-	dev, err := ns.devices.SetByID(ctx, req.Device.EndDeviceIdentifiers.ApplicationIdentifiers, req.Device.EndDeviceIdentifiers.DeviceID, req.FieldMask.Paths, func(dev *ttnpb.EndDevice) (*ttnpb.EndDevice, []string, error) {
-		paths := req.FieldMask.Paths
+	dev, err := ns.devices.SetByID(ctx, req.Device.EndDeviceIdentifiers.ApplicationIdentifiers, req.Device.EndDeviceIdentifiers.DeviceID, gets, func(dev *ttnpb.EndDevice) (*ttnpb.EndDevice, []string, error) {
+		if err := ttnpb.ProhibitFields(req.FieldMask.Paths,
+			"mac_state.current_parameters",
+			"mac_state.desired_parameters",
+			"pending_session",
+		); err != nil {
+			return nil, nil, errInvalidFieldMask.WithCause(err)
+		}
+
 		if dev != nil {
-			addDownlinkTask = ttnpb.HasAnyField(paths, "mac_state.device_class") && req.Device.MACState.DeviceClass != ttnpb.CLASS_A ||
-				ttnpb.HasAnyField(paths, "queued_application_downlinks") && len(req.Device.QueuedApplicationDownlinks) > 0
-			return &req.Device, paths, nil
+			if err := ttnpb.ProhibitFields(req.FieldMask.Paths,
+				"session",
+			); err != nil {
+				return nil, nil, errInvalidFieldMask.WithCause(err)
+			}
+
+			addDownlinkTask = ttnpb.HasAnyField(req.FieldMask.Paths,
+				"mac_state.device_class",
+				"queued_application_downlinks",
+			) && req.Device.MACState.DeviceClass != ttnpb.CLASS_A &&
+				(len(req.Device.QueuedApplicationDownlinks) > 0 || !req.Device.MACState.CurrentParameters.Equal(req.Device.MACState.DesiredParameters))
```

Comment:
Please make this statement readable

Reply:
What would be "readable" in this case and what makes this not readable?
Can you make a suggestion?(ctrl+g)

Reply:
That, or variables and a nice logical expression.

Reply:
But I'm also fine leaving it this way.